### ISSUE TITLE
Use Joi for validation

### ISSUE THREAD

**Issue by mattapperson** [balderdashy/waterline#824]
Waterline is a great project, but the validation in models is a little weak. Had or would consideration be given to moving to https://github.com/hapijs/joi for this aspect of the project?

**Comment by particlebanana** [comment]
I've briefly looked into Joi as a replacement for Anchor but haven't evaluated how much effort it would involve or if the api would still work as expected. Would love to get it moved over though!

**Comment by mattapperson** [comment]
ill take a look and see what I / my team can do :)

**Comment by kevinburkeshyp** [comment]
maybe could try relying on your database for validation? for example with
not null constraints, or in postgres you can do CHECK positive_number
(amount > 0). pretty sure range constraints are pretty easy as well. that
way bad values can't sneak in.

On Wed, Feb 4, 2015 at 1:32 PM, Matt Apperson <notifications@github.com>
wrote:

> ill take a look and see what I / my team can do :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balderdashy/waterline/issues/824#issuecomment-72943929>
> .
>



-- 
kevin

**Comment by mattapperson** [comment]
That only works on some databases, nosql solutions don't have that concept

**Comment by tjwebb** [comment]
> maybe could try relying on your database for validation?

Then you have to re-implement validation for each database. This defeats the purpose of using an ORM layer in the first place.

**Comment by mattapperson** [comment]
@tjwebb beyond the scope of what I feel like building (as I dont need it at all...) but perhaps we could provide a hook for adaptors to plug into to make sure validation matches?

**Comment by kevinburkeshyp** [comment]
I'm confused. As a consumer of an ORM, I'm using one, maybe two database
technologies.

I'm also (probably) using multiple different languages to access a
database, which is why it's better to have these checks in the database and
not in the application, where they can be bypassed.

On Wed, Feb 4, 2015 at 2:46 PM, Travis Webb <notifications@github.com>
wrote:

> maybe could try relying on your database for validation?
>
> Then you have to re-implement validation for each database. This defeats
> the purpose of using an ORM layer in the first place.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balderdashy/waterline/issues/824#issuecomment-72956821>
> .
>



-- 
kevin

**Comment by mattapperson** [comment]
@kevinburkeshyp you are assuming that all DBs implement validation. That is an incorrect assumption.

**Comment by tjwebb** [comment]
> @tjwebb beyond the scope of what I feel like building

I don't know what you mean. I didn't suggest building anything.

**Comment by mattapperson** [comment]
@tjwebb Right... I am suggesting we provide a hook though so that if the developer of an adaptor wanted to, they could ensure that the apps designated validation did not conflict with the DB... but only if the adaptor dev WANTED to.

**Comment by tjwebb** [comment]
> I'm using one, maybe two database technologies.

If you're willing to bet on never needing to integrate with another database vendor for the lifetime of your project, then just write SQL through your database driver. If not, I'd follow the design patterns of your chosen framework rather than doing things your way because you think its better. 

> which is why it's better to have these checks in the database and not in the application, where they can be bypassed.

That's governed by your system design. Most legacy systems place the database at the center of the universe. Most modern systems place web middleware at the center as the main integration point through which all database interaction occurs.

**Comment by mattapperson** [comment]
@tjwebb well put

**Comment by tjwebb** [comment]
> they could ensure that the apps designated validation did not conflict with the DB

That would work between the adapter and database X, but when you switch to database Y, that means your validation rules specific to database X might conflict. In the context of Waterline, where you want to define one ORM for any database, this would be problematic. That's why we use a single validation library.

If the discussion is should we replace that *one* library with a different one, I'm open to talking about that.

**Comment by mattapperson** [comment]
All I care about is Joi, I have zero need/want for DB validation

Was just trying to make everyone happy :)

**Comment by tjwebb** [comment]
Yea; I've got another issue open seeking to replace waterline-sequel with knex.js. It's better supported and more robust. I think your suggestion is a good one.

**Comment by particlebanana** [comment]
The feature request here is to replace [Anchor](https://github.com/balderdashy/anchor), our validation library, with [Joi](https://github.com/hapijs/joi). Joi looks nice and seems to be well supported so as long as it can work with the current validation api it sounds good to me.

@kevinburkeshyp in reply to your concerns about the database level constraints you should absolutely follow the normal database design best practices as far as unique, defaults, required, etc constraints are concerned! The app level validation libraries allow you to handle nice app level stuff like validating email address, etc without the overhead of a network or database request. It's equivalent to something like Active Record validations or Django validators.

**Comment by mattapperson** [comment]
I think for this it would be best to support the current API so as to not break compatibility, but also let devs use the direct Joi API

**Comment by mattapperson** [comment]
@particlebanana Is this ok with you? ˆˆ

**Comment by devinivy** [comment]
+1, Joi is wonderful!  This integration would be relatively simple.

**Comment by g-div** [comment]
+1 Joi seems to be pretty reliable as validation library. It's widely used in production and would be nice to have such stability under the hood of the Waterline's ORM functions.

**Comment by darinc** [comment]
+1 for using Joi, it's a great validator

**Comment by dmarcelino** [comment]
Should we create some sort of `waterline-joi` or `waterline-validation` project that could be used as a drop in replacement for `anchor`, fulfilling the bellow requirements:

1. implements the current validation API (passes same tests as `anchor`, etc);
2. extend current set of validation with `joi` validations.

We've closed `0.10` so now we'll be focusing on `0.11` features. Any volunteers?

**Comment by davemackintosh** [comment]
Pretty rudimentary for now and it's "in" another project for now but I do plan to take this out and build it out a bit more. Hopefully it helps!

https://github.com/newworldcode/rainbow/blob/master/blueprint-to-joi.js

**Comment by Caligone** [comment]
+1 for using Joi ! Frequently maintened and good support
The @dmarcelino 's waterline-validation idea seems great.
I could help a bit

**Comment by sailsbot** [comment]
Thanks for posting, @mattapperson.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by davemackintosh** [comment]
Much better version, still not 100% but it's mostly there.

https://www.npmjs.com/package/waterline-joi
https://github.com/newworldcode/waterline-joi

**Comment by devinivy** [comment]
Nice, @davemackintosh!

**Comment by davemackintosh** [comment]
Thanks @devinivy, it's not perfect but it's much more feature packed and tidier than my original.
